### ISSUE TITLE
needless_match doesn't consider the `else` of `if let`

### ISSUE THREAD

**Issue by goffrie** [rust-lang/rust-clippy#8695]
### Summary

The `needless_match` lint produces a nonsensical suggestion in some cases. It seems like it fires entirely based on the `if`-clause of an `if let`, ignoring the `else` branch's value entirely. This is a regression, most likely from #8549 (but I haven't confirmed that).

### Lint Name

needless_match

### Reproducer

[I tried this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=db9096fe9508b1f92d94b985ec04996a):

```rust
pub fn bad() -> Option<i32> {
    if let Some(existing) = None {
        Some(existing)
    } else {
        Some(123)
    }
}
```

I saw this happen:

```
warning: this if-let expression is unnecessary
 [--> src/lib.rs:2:5
](https://play.rust-lang.org/#)  |
2 | /     if let Some(existing) = None {
3 | |         Some(existing)
4 | |     } else {
5 | |         Some(123)
6 | |     }
  | |_____^ help: replace it with: `None`
  |
  = note: `#[warn(clippy::needless_match)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_match
```

I expected to see no lint error.

### Version

```text
rustc 1.62.0-nightly (1f7fb6413 2022-04-10)
binary: rustc
commit-hash: 1f7fb6413d6d6c0c929b223e478e44c3db991b03
commit-date: 2022-04-10
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.0
```

**Comment by fanninpm** [comment]
Does #8700 fix this?

**Comment by goffrie** [comment]
> Does #8700 fix this?

Yes, I think it does.
